python/shotgun_data/sgdata.py

```diff
@@ -594,61 +613,64 @@ def run(self):
-            # condition with clients wanting to set the connection.
-            #
-            # The problem arises when:
-            # User instantiates the SgDataRetriever
-            # Launches the thread.
-            # Set's the connection.
-            #
-            # Since the thread also sets the connection, there's a race condition.
-            #
-            # This way of using the SgDataRetriever is kinda eroneous. The user
-            # should probably create the retriever, set the connection and then
-            # launch the thread. However, existing code is sometimes hit by this
-            # race condition because the class is being used the wrong way so it's
-            # safer to just update this class for now.
-            if self.__sg is None:
-                # create our own private shotgun connection. This is because
-                # the shotgun API isn't threadsafe, so running multiple models in parallel
-                # (common) may result in side effects if a single connection is shared
-                self.__sg = tank.util.shotgun.create_sg_connection()
-
-                # set the maximum timeout for this connection for fluency
-                self.__sg.config.timeout_secs = CONNECTION_TIMEOUT_SECS
-
             # Step 2. Process next item and send signals.
             try:
 
                 # process the item:
+                if item_type == ShotgunDataRetriever._THUMB_CHECK:
+                    # check if a thumbnail exists on disk. If not, fall back onto
+                    # a thumbnail download from shotgun/s3
+                    url = item_to_process["url"]
+                    path_to_cached_thumb = self._get_thumbnail_path(url, self._bundle)
+                    if os.path.exists(path_to_cached_thumb):
+                        # thumbnail already here! yay!
+                        if item_to_process["load_image"]:
+                            image = QtGui.QImage()
+                            image.load(path_to_cached_thumb)
+                        else:
+                            image = None
+                        self.work_completed.emit(item_to_process["id"], 
+                                                 "thumb", 
+                                                 {"thumb_path": path_to_cached_thumb, "image": image} )
+                    else:
+                        # no thumb here. Stick the data into the thumb download queue to request download
+                        self._queue_mutex.lock()
+                        try:
+                            self._thumb_download_queue.append(item_to_process)
+                        finally:
+                            self._queue_mutex.unlock()
 
-                if item_type == ShotgunDataRetriever._SG_FIND_QUERY:
+                elif item_type == ShotgunDataRetriever._SG_FIND_QUERY:
                     # get stuff from shotgun
+                    self.__ensure_sg_connection_initialized()
```

Comment:
Definitely have a property that lazily initializes __sg feels the right solution. This is getting copy pasted all over the place and could easily be missed in a branch.